File: backend/src/requests/requests.controller.ts
import {
  Controller,
  Post,
  Get,
  Body,
  Param,
  Query,
  UseGuards,
  Request,
} from '@nestjs/common';
import { ApiTags, ApiOperation, ApiResponse, ApiParam, ApiQuery, ApiBody } from '@nestjs/swagger';
import { RequestsService } from './requests.service';
import { CreateRequestDto } from './dto/create-request.dto';
import { CreateQuoteDto } from './dto/create-quote.dto';

// Simple auth guard - in production, use proper JWT auth
class SimpleAuthGuard {
  canActivate(context: any) {
    // For now, extract userId from request headers or body
    // In production, use JWT token
    return true;
  }
}

@ApiTags('requests')
@Controller('api/requests')
export class RequestsController {
  constructor(private readonly requestsService: RequestsService) {}

  @Post('create')
  @ApiOperation({ summary: 'Create a new fuel request' })
  @ApiResponse({ status: 201, description: 'Request created successfully' })
  @ApiResponse({ status: 400, description: 'Invalid input data' })
  @ApiBody({ type: CreateRequestDto })
  async createRequest(@Body() createDto: CreateRequestDto) {
    if (!createDto.userId) {
      return {
        error: 'User ID is required',
        message: 'Please provide a valid user ID',
      };
    }

    // Get actual user role from location service to preserve it
    const locationService = this.requestsService['locationService'];
    const existingLocation = locationService.getLocation(createDto.userId);
    const actualUserRole = existingLocation?.role || createDto.userRole || 'needy';
    
    // Request role defaults to user's actual role if not specified
    const requestRole = createDto.userRole || actualUserRole;
    
    console.log(`📝 Creating request:`, {
      userId: createDto.userId,
      actualUserRole: actualUserRole,
      requestRole: requestRole,
      location: { lat: createDto.latitude, lon: createDto.longitude },
      message: createDto.message,
    });
    
    const request = await this.requestsService.createRequest(
      createDto.userId,
      createDto.latitude,
      createDto.longitude,
      createDto.message,
      createDto.quantityLiters,
      createDto.urgency || 'normal',
      requestRole, // Pass request role (who created it)
    );
    
    console.log('✅ Request created:', request.id, 'by', requestRole, createDto.userId, `(user's role: ${actualUserRole})`);
    return {
      ...request,
      message: `Request created successfully. It will be visible to nearby ${requestRole === 'needy' ? 'providers' : 'needers'}.`,
    };
  }

  @Get('nearest')
  @ApiOperation({ summary: 'Find nearest fuel requests' })
  @ApiQuery({ name: 'latitude', type: Number, description: 'Latitude of the search location' })
  @ApiQuery({ name: 'longitude', type: Number, description: 'Longitude of the search location' })
  @ApiQuery({ name: 'maxDistance', type: Number, required: false, description: 'Maximum distance in km (default: 10)' })
  @ApiQuery({ name: 'limit', type: Number, required: false, description: 'Maximum number of results (default: 10)' })
  @ApiQuery({ name: 'providerId', type: String, required: false, description: 'Provider user ID' })
  @ApiQuery({ name: 'needyId', type: String, required: false, description: 'Needy user ID' })
  @ApiQuery({ name: 'userRole', enum: ['needy', 'provider'], required: false, description: 'User role' })
  @ApiResponse({ status: 200, description: 'List of nearest requests' })
  async findNearestRequests(
    @Query('latitude') latitude: number,
    @Query('longitude') longitude: number,
    @Query('maxDistance') maxDistance: number = 10,
    @Query('limit') limit: number = 10,
    @Query('providerId') providerId: string, // Can be provider ID or needy ID
    @Query('needyId') needyId?: string, // Optional: if provided, include needy's own requests
    @Query('userRole') userRole?: 'needy' | 'provider', // User role to determine what to show
  ) {
    // Use needyId if provided, otherwise use providerId
    const userId = needyId || providerId;
    if (!userId) {
      return {
        error: 'User ID is required (providerId or needyId)',
        requests: [],
        count: 0,
      };
    }

    console.log('🔍 Searching for requests:', {
      userId: userId,
      userRole: userRole,
      isNeedy: !!needyId,
      realLocation: { latitude: Number(latitude), longitude: Number(longitude) },
      maxDistance: Number(maxDistance),
    });

    // Register user location - PRESERVE existing role!
    const locationService = this.requestsService['locationService'];
    const existingLocation = locationService.getLocation(userId);
    
    // IMPORTANT: Preserve the user's actual role, don't hardcode it
    const actualRole = existingLocation?.role || userRole || (needyId ? 'needy' : 'provider');
    
    locationService.updateLocation(userId, {
      userId: userId,
      name: existingLocation?.name || (needyId ? 'Needy User' : 'Provider'), // Use existing name or default
      role: actualRole, // Preserve existing role, don't change it
      latitude: Number(latitude),
      longitude: Number(longitude),
      isAvailable: true,
    });
    // Don't log routine location updates - too verbose

    // Pass userId and userRole to include relevant requests (pending + accepted for providers)
    // Use actualRole determined from location service, fallback to userRole parameter
    const finalUserRole = actualRole as 'needy' | 'provider';
    console.log(`🔍 Final userRole for findNearestRequests: ${finalUserRole} (from location: ${existingLocation?.role}, param: ${userRole})`);
    const requests = await this.requestsService.findNearestRequests(
      Number(latitude),
      Number(longitude),
      Number(maxDistance),
      Number(limit),
      userId, // Pass userId for both needy and provider
      finalUserRole, // Pass determined user role
    );
    
    console.log(`✅ Found ${requests.length} requests nearby`);
    return {
      requests,
      count: requests.length,
      searchLocation: { latitude: Number(latitude), longitude: Number(longitude) },
    };
  }

  @Get('providers/nearest')
  async findNearestProviders(
    @Query('latitude') latitude: number,
    @Query('longitude') longitude: number,
    @Query('maxDistance') maxDistance: number = 10,
    @Query('limit') limit: number = 10,
    @Query('needyId') needyId: string, // Required - real needy ID
  ) {
    if (!needyId) {
      return {
        error: 'Needy ID is required',
        providers: [],
        count: 0,
      };
    }

    console.log('🔍 Real needy searching for providers:', {
      needyId: needyId,
      realLocation: { latitude: Number(latitude), longitude: Number(longitude) },
      maxDistance: Number(maxDistance),
    });

    // Register real needy location - PRESERVE existing role!
    const locationService = this.requestsService['locationService'];
    const existingLocation = locationService.getLocation(needyId);
    
    // IMPORTANT: Preserve the user's actual role, don't hardcode it
    const actualRole = existingLocation?.role || 'needy';
    
    locationService.updateLocation(needyId, {
      userId: needyId,
      name: existingLocation?.name || 'Needy User', // Use existing name or default
      role: actualRole, // Preserve existing role, don't change it
      latitude: Number(latitude),
      longitude: Number(longitude),
      isAvailable: true,
    });
    // Don't log routine location updates - too verbose

    const providers = this.requestsService.findNearestProviders(
      Number(latitude),
      Number(longitude),
      Number(maxDistance),
      Number(limit),
      needyId, // Pass needyId to exclude themselves from results
    );
    
    console.log(`✅ Found ${providers.length} providers nearby`);
    return {
      providers,
      count: providers.length,
      searchLocation: { latitude: Number(latitude), longitude: Number(longitude) },
    };
  }

  @Get('needers/nearest')
  async findNearestNeeders(
    @Query('latitude') latitude: number,
    @Query('longitude') longitude: number,
    @Query('maxDistance') maxDistance: number = 10,
    @Query('limit') limit: number = 10,
    @Query('providerId') providerId: string, // Required - real provider ID
  ) {
    if (!providerId) {
      return {
        error: 'Provider ID is required',
        needers: [],
        count: 0,
      };
    }

    console.log('🔍 Real provider searching for needers:', {
      providerId: providerId,
      realLocation: { latitude: Number(latitude), longitude: Number(longitude) },
      maxDistance: Number(maxDistance),
    });

    // Register real provider location - PRESERVE existing role!
    const locationService = this.requestsService['locationService'];
    const existingLocation = locationService.getLocation(providerId);
    
    // IMPORTANT: Preserve the user's actual role, don't hardcode it
    const actualRole = existingLocation?.role || 'provider';
    
    locationService.updateLocation(providerId, {
      userId: providerId,
      name: existingLocation?.name || 'Provider',
      role: actualRole, // Preserve existing role, don't change it
      latitude: Number(latitude),
      longitude: Number(longitude),
      isAvailable: true,
    });
    // Don't log routine location updates - too verbose

    const needers = this.requestsService.findNearestNeeders(
      Number(latitude),
      Number(longitude),
      Number(maxDistance),
      Number(limit),
      providerId, // Pass providerId to exclude themselves from results
    );
    
    console.log(`✅ Found ${needers.length} needers nearby`);
    return {
      needers,
      count: needers.length,
      searchLocation: { latitude: Number(latitude), longitude: Number(longitude) },
    };
  }

  @Post(':id/accept')
  @ApiOperation({ summary: 'Accept a fuel request' })
  @ApiParam({ name: 'id', description: 'Request ID' })
  @ApiResponse({ status: 200, description: 'Request accepted successfully' })
  @ApiResponse({ status: 404, description: 'Request not found' })
  async acceptRequest(
    @Param('id') requestId: string,
    @Body('providerId') providerId: string,
  ) {
    const request = await this.requestsService.acceptRequest(requestId, providerId);
    return request;
  }

  /**
   * Get request history (completed requests)
   * IMPORTANT: This route MUST come before @Get(':id') to avoid route conflicts
   */
  @Get('history')
  @ApiOperation({ summary: 'Get request history (completed requests)' })
  @ApiQuery({ name: 'userId', type: String, description: 'User ID' })
  @ApiQuery({ name: 'userRole', enum: ['needy', 'provider'], description: 'User role' })
  @ApiResponse({ status: 200, description: 'Request history retrieved successfully' })
  async getRequestHistory(
    @Query('userId') userId: string,
    @Query('userRole') userRole: 'needy' | 'provider',
  ) {
    console.log(`\n\n📜 ========== HISTORY ENDPOINT CALLED ==========`);
    console.log(`📜 History endpoint called: userId=${userId}, userRole=${userRole}`);
    console.log(`📜 Query params received:`, { userId, userRole });
    console.log(`📜 Request received at: ${new Date().toISOString()}`);
    
    if (!userId || !userRole) {
      console.error('❌ Missing userId or userRole');
      console.error('❌ userId:', userId, 'userRole:', userRole);
      return {
        error: 'User ID and role are required',
        message: 'Please provide userId and userRole',
        history: [],
        count: 0,
      };
    }

    try {
      console.log(`📜 Calling getRequestHistory service method...`);
      const history = await this.requestsService.getRequestHistory(userId, userRole);
      console.log(`📜 Service returned ${history.length} history items`);
      console.log(`📜 History items:`, history.map(h => ({
        id: h.id,
        status: h.status,
        needyId: h.needyId,
        acceptedBy: h.acceptedBy,
      })));
      return {
        success: true,
        history,
        count: history.length,
      };
    } catch (error: any) {
      console.error('❌ Error getting history:', error);
      console.error('❌ Error stack:', error.stack);
      return {
        success: false,
        error: error.message || 'Failed to get history',
        history: [],
        count: 0,
      };
    }
  }

  /**
   * Get active requests (pending, accepted, in_progress)
   * IMPORTANT: This route MUST come before @Get(':id') to avoid route conflicts
   */
  @Get('active')
  async getActiveRequests(
    @Query('userId') userId: string,
    @Query('userRole') userRole: 'needy' | 'provider',
  ) {
    if (!userId || !userRole) {
      return {
        error: 'User ID and role are required',
        message: 'Please provide userId and userRole',
      };
    }

    const activeRequests = await this.requestsService.getActiveRequests(userId, userRole);
    return {
      success: true,
      requests: activeRequests,
      count: activeRequests.length,
    };
  }

  @Get(':id')
  @ApiOperation({ summary: 'Get a specific request by ID' })
  @ApiParam({ name: 'id', description: 'Request ID' })
  @ApiResponse({ status: 200, description: 'Request found' })
  @ApiResponse({ status: 404, description: 'Request not found' })
  async getRequest(@Param('id') requestId: string) {
    const request = await this.requestsService.getRequest(requestId);
    if (!request) {
      return { error: 'Request not found' };
    }
    return request;
  }

  @Get('user/:userId')
  async getUserRequests(@Param('userId') userId: string) {
    const requests = await this.requestsService.getUserRequests(userId);
    return { requests };
  }

  @Post(':id/cancel')
  async cancelRequest(
    @Param('id') requestId: string,
    @Body('userId') userId: string,
  ) {
    const request = await this.requestsService.cancelRequest(requestId, userId);
    return request;
  }

  @Post(':id/complete')
  async completeRequest(
    @Param('id') requestId: string,
    @Body('userId') userId: string,
    @Body('userRole') userRole: 'needy' | 'provider',
    @Body('providerId') providerId?: string, // Backward compatibility
  ) {
    // Support both old and new API formats
    const finalUserId = userId || providerId;
    const finalUserRole = userRole || (providerId ? 'provider' : 'needy');
    
    if (!finalUserId) {
      return {
        error: 'User ID is required',
        message: 'Please provide userId (or providerId for backward compatibility)',
      };
    }

    try {
      const request = await this.requestsService.completeRequest(requestId, finalUserId, finalUserRole);
      return {
        success: true,
        request: {
          id: request.id,
          status: request.status,
          completedAt: request.updatedAt,
        },
        message: 'Request completed successfully. Moved to history.',
      };
    } catch (error: any) {
      return {
        error: error.message || 'Failed to complete request',
        message: error.message || 'An error occurred while completing the request',
      };
    }
  }

  // Quote Endpoints
  @Post('quotes/create')
  @ApiOperation({ summary: 'Create a quote for a fuel request' })
  @ApiResponse({ status: 201, description: 'Quote created successfully' })
  @ApiResponse({ status: 400, description: 'Invalid input data' })
  @ApiBody({ type: CreateQuoteDto })
  async createQuote(@Body() createQuoteDto: CreateQuoteDto) {
    if (!createQuoteDto.providerId) {
      return {
        error: 'Provider ID is required',
        message: 'Please provide a valid provider ID',
      };
    }

    if (!createQuoteDto.requestId) {
      return {
        error: 'Request ID is required',
        message: 'Please provide a valid request ID',
      };
    }

    console.log('💰 Creating quote:', {
      requestId: createQuoteDto.requestId,
      providerId: createQuoteDto.providerId,
      price: createQuoteDto.price,
      currency: createQuoteDto.currency,
      estimatedDeliveryTime: createQuoteDto.estimatedDeliveryTime,
    });

    try {
      const quote = await this.requestsService.createQuote(
        createQuoteDto.requestId,
        createQuoteDto.providerId,
        createQuoteDto.price,
        createQuoteDto.currency,
        createQuoteDto.estimatedDeliveryTime,
        createQuoteDto.message,
      );
      console.log('✅ Quote created successfully:', quote.id);
      return {
        ...quote,
        message: 'Quote sent successfully to needy user',
      };
    } catch (error: any) {
      console.error('❌ Error creating quote:', error);
      return {
        error: error.message || 'Failed to create quote',
        message: error.message || 'An error occurred while creating the quote',
      };
    }
  }

  @Get('quotes/request/:requestId')
  async getQuotesForRequest(@Param('requestId') requestId: string) {
    const quotes = await this.requestsService.getQuotesForRequest(requestId);
    return {
      quotes,
      count: quotes.length,
    };
  }

  @Get('quotes/needy/:needyId')
  async getQuotesForNeedy(@Param('needyId') needyId: string) {
    const quotes = await this.requestsService.getQuotesForNeedy(needyId);
    return {
      quotes,
      count: quotes.length,
    };
  }

  @Get('quotes/provider/:providerId')
  async getQuotesForProvider(@Param('providerId') providerId: string) {
    const quotes = await this.requestsService.getQuotesForProvider(providerId);
    return {
      quotes,
      count: quotes.length,
    };
  }

  @Post('quotes/:quoteId/accept')
  @ApiOperation({ summary: 'Accept a quote' })
  @ApiParam({ name: 'quoteId', description: 'Quote ID' })
  @ApiResponse({ status: 200, description: 'Quote accepted successfully' })
  @ApiResponse({ status: 404, description: 'Quote not found' })
  async acceptQuote(
    @Param('quoteId') quoteId: string,
    @Body('needyId') needyId: string,
  ) {
    try {
      console.log('✅ Accepting quote:', { quoteId, needyId });
      
      if (!needyId || needyId.trim() === '') {
        console.error('❌ Needy ID is missing or empty');
        throw new Error('Needy ID is required');
      }

      const quote = await this.requestsService.acceptQuote(quoteId, needyId);
      console.log('✅ Quote accepted successfully:', quote.id);
      
      const request = await this.requestsService.getRequest(quote.requestId);
      if (!request) {
        console.error(`❌ Request ${quote.requestId} not found after accepting quote`);
        // Still return success since quote was accepted, but warn about missing request
        return {
          ...quote,
          request: null,
          warning: 'Request not found',
          message: 'Quote accepted successfully, but request details could not be retrieved.',
        };
      }
      
      console.log('✅ Request retrieved:', { id: request.id, status: request.status, needyId: request.needyId, acceptedBy: request.acceptedBy });
      
      return {
        ...quote,
        request: {
          id: request.id,
          status: request.status,
          needyId: request.needyId,
          acceptedBy: request.acceptedBy,
        },
        message: 'Quote accepted successfully. Request is now hidden from other providers.',
      };
    } catch (error: any) {
      console.error('❌ Error accepting quote:', error);
      console.error('Error details:', {
        message: error.message,
        stack: error.stack,
        quoteId,
        needyId,
      });
      throw error; // Let NestJS handle the error response
    }
  }

  @Post('quotes/:quoteId/reject')
  async rejectQuote(
    @Param('quoteId') quoteId: string,
    @Body('needyId') needyId: string,
  ) {
    const quote = await this.requestsService.rejectQuote(quoteId, needyId);
    return {
      ...quote,
      message: 'Quote rejected',
    };
  }

  @Get('quotes/:quoteId')
  async getQuote(@Param('quoteId') quoteId: string) {
    const quote = await this.requestsService.getQuote(quoteId);
    if (!quote) {
      return { error: 'Quote not found' };
    }
    return quote;
  }

}

File: backend-dotnet/Controllers/RequestsController.cs
using Microsoft.AspNetCore.Mvc;
using FuelMateBackend.Services;
using FuelMateBackend.DTOs;
using FuelMateBackend.Models;

namespace FuelMateBackend.Controllers;

[ApiController]
[Route("api/[controller]")]
[Tags("requests")]
public class RequestsController : ControllerBase
{
    private readonly RequestsService _requestsService;
    private readonly LocationService _locationService;

    public RequestsController(RequestsService requestsService, LocationService locationService)
    {
        _requestsService = requestsService;
        _locationService = locationService;
    }

    /// <summary>
    /// Create a new fuel request
    /// </summary>
    [HttpPost("create")]
    public async Task<IActionResult> CreateRequest([FromBody] CreateRequestDto dto)
    {
        if (string.IsNullOrEmpty(dto.UserId))
        {
            return BadRequest(new
            {
                error = "User ID is required",
                message = "Please provide a valid user ID"
            });
        }

        var existingLocation = _locationService.GetLocation(dto.UserId);
        var actualUserRole = existingLocation?.Role ?? dto.UserRole ?? "needy";
        var requestRole = dto.UserRole ?? actualUserRole;

        Console.WriteLine($"📝 Creating request: userId={dto.UserId}, actualUserRole={actualUserRole}, requestRole={requestRole}, location=({dto.Latitude},{dto.Longitude})");

        var request = await _requestsService.CreateRequest(
            dto.UserId,
            dto.Latitude,
            dto.Longitude,
            dto.Message,
            dto.QuantityLiters,
            dto.Urgency,
            requestRole);

        Console.WriteLine($"✅ Request created: {request.Id} by {requestRole} {dto.UserId} (user's role: {actualUserRole})");

        return Ok(new
        {
            request.Id,
            request.NeedyId,
            request.Role,
            request.Latitude,
            request.Longitude,
            request.Message,
            request.QuantityLiters,
            request.Urgency,
            request.Status,
            request.CreatedAt,
            successMessage = $"Request created successfully. It will be visible to nearby {(requestRole == "needy" ? "providers" : "needers")}."
        });
    }

    /// <summary>
    /// Find nearest fuel requests
    /// </summary>
    [HttpGet("nearest")]
    public async Task<IActionResult> FindNearestRequests(
        [FromQuery] decimal latitude,
        [FromQuery] decimal longitude,
        [FromQuery] double maxDistance = 10,
        [FromQuery] int limit = 10,
        [FromQuery] string? providerId = null,
        [FromQuery] string? needyId = null,
        [FromQuery] string? userRole = null)
    {
        var userId = needyId ?? providerId;
        if (string.IsNullOrEmpty(userId))
        {
            return BadRequest(new
            {
                error = "User ID is required (providerId or needyId)",
                requests = new List<object>(),
                count = 0
            });
        }

        Console.WriteLine($"🔍 Searching for requests: userId={userId}, userRole={userRole}, location=({latitude},{longitude}), maxDistance={maxDistance}");

        // Register/update user location
        var existingLocation = _locationService.GetLocation(userId);
        var actualRole = existingLocation?.Role ?? userRole ?? (needyId != null ? "needy" : "provider");

        _locationService.UpdateLocation(userId, new UserLocation
        {
            UserId = userId,
            Name = existingLocation?.Name ?? (needyId != null ? "Needy User" : "Provider"),
            Role = actualRole,
            Latitude = latitude,
            Longitude = longitude,
            IsAvailable = true
        });

        var requests = await _requestsService.FindNearestRequests(
            latitude,
            longitude,
            maxDistance,
            limit,
            userId,
            actualRole);

        Console.WriteLine($"✅ Found {requests.Count} requests nearby");

        return Ok(new
        {
            requests,
            count = requests.Count,
            searchLocation = new { latitude, longitude }
        });
    }

    /// <summary>
    /// Find nearest providers
    /// </summary>
    [HttpGet("providers/nearest")]
    public IActionResult FindNearestProviders(
        [FromQuery] decimal latitude,
        [FromQuery] decimal longitude,
        [FromQuery] double maxDistance = 10,
        [FromQuery] int limit = 10,
        [FromQuery] string? needyId = null)
    {
        if (string.IsNullOrEmpty(needyId))
        {
            return BadRequest(new
            {
                error = "Needy ID is required",
                providers = new List<object>(),
                count = 0
            });
        }

        Console.WriteLine($"🔍 Real needy searching for providers: needyId={needyId}, location=({latitude},{longitude})");

        // Register/update needy location
        var existingLocation = _locationService.GetLocation(needyId);
        _locationService.UpdateLocation(needyId, new UserLocation
        {
            UserId = needyId,
            Name = existingLocation?.Name ?? "Needy User",
            Role = existingLocation?.Role ?? "needy",
            Latitude = latitude,
            Longitude = longitude,
            IsAvailable = true
        });

        var providers = _requestsService.FindNearestProviders(latitude, longitude, maxDistance, limit, needyId);

        Console.WriteLine($"✅ Found {providers.Count} providers nearby");

        return Ok(new
        {
            providers,
            count = providers.Count,
            searchLocation = new { latitude, longitude }
        });
    }

    /// <summary>
    /// Find nearest needers
    /// </summary>
    [HttpGet("needers/nearest")]
    public IActionResult FindNearestNeeders(
        [FromQuery] decimal latitude,
        [FromQuery] decimal longitude,
        [FromQuery] double maxDistance = 10,
        [FromQuery] int limit = 10,
        [FromQuery] string? providerId = null)
    {
        if (string.IsNullOrEmpty(providerId))
        {
            return BadRequest(new
            {
                error = "Provider ID is required",
                needers = new List<object>(),
                count = 0
            });
        }

        Console.WriteLine($"🔍 Real provider searching for needers: providerId={providerId}, location=({latitude},{longitude})");

        // Register/update provider location
        var existingLocation = _locationService.GetLocation(providerId);
        _locationService.UpdateLocation(providerId, new UserLocation
        {
            UserId = providerId,
            Name = existingLocation?.Name ?? "Provider",
            Role = existingLocation?.Role ?? "provider",
            Latitude = latitude,
            Longitude = longitude,
            IsAvailable = true
        });

        var needers = _requestsService.FindNearestNeeders(latitude, longitude, maxDistance, limit, providerId);

        Console.WriteLine($"✅ Found {needers.Count} needers nearby");

        return Ok(new
        {
            needers,
            count = needers.Count,
            searchLocation = new { latitude, longitude }
        });
    }

    /// <summary>
    /// Accept a fuel request
    /// </summary>
    [HttpPost("{id}/accept")]
    public async Task<IActionResult> AcceptRequest(string id, [FromBody] Dictionary<string, string> body)
    {
        if (!body.TryGetValue("providerId", out var providerId))
        {
            return BadRequest("Provider ID is required");
        }

        var request = await _requestsService.AcceptRequest(id, providerId);
        if (request == null)
        {
            return NotFound(new { error = "Request not found" });
        }

        return Ok(request);
    }

    /// <summary>
    /// Get request history (completed requests)
    /// </summary>
    [HttpGet("history")]
    public async Task<IActionResult> GetRequestHistory(
        [FromQuery] string? userId,
        [FromQuery] string? userRole)
    {
        Console.WriteLine($"\n\n📜 ========== HISTORY ENDPOINT CALLED ==========");
        Console.WriteLine($"📜 History endpoint called: userId={userId}, userRole={userRole}");

        if (string.IsNullOrEmpty(userId) || string.IsNullOrEmpty(userRole))
        {
            Console.WriteLine("❌ Missing userId or userRole");
            return BadRequest(new
            {
                error = "User ID and role are required",
                message = "Please provide userId and userRole",
                history = new List<object>(),
                count = 0
            });
        }

        try
        {
            Console.WriteLine("📜 Calling getRequestHistory service method...");
            var history = await _requestsService.GetRequestHistory(userId, userRole);
            Console.WriteLine($"📜 Service returned {history.Count} history items");

            return Ok(new
            {
                success = true,
                history,
                count = history.Count
            });
        }
        catch (Exception ex)
        {
            Console.WriteLine($"❌ Error getting history: {ex.Message}");
            return StatusCode(500, new
            {
                success = false,
                error = ex.Message,
                history = new List<object>(),
                count = 0
            });
        }
    }

    /// <summary>
    /// Get active requests (pending, accepted, in_progress)
    /// </summary>
    [HttpGet("active")]
    public async Task<IActionResult> GetActiveRequests(
        [FromQuery] string? userId,
        [FromQuery] string? userRole)
    {
        if (string.IsNullOrEmpty(userId) || string.IsNullOrEmpty(userRole))
        {
            return BadRequest(new
            {
                error = "User ID and role are required",
                message = "Please provide userId and userRole"
            });
        }

        var activeRequests = await _requestsService.GetActiveRequests(userId, userRole);

        return Ok(new
        {
            success = true,
            requests = activeRequests,
            count = activeRequests.Count
        });
    }

    /// <summary>
    /// Get a specific request by ID
    /// </summary>
    [HttpGet("{id}")]
    public async Task<IActionResult> GetRequest(string id)
    {
        var request = await _requestsService.GetRequest(id);
        if (request == null)
        {
            return NotFound(new { error = "Request not found" });
        }

        return Ok(request);
    }

    /// <summary>
    /// Get all requests for a user
    /// </summary>
    [HttpGet("user/{userId}")]
    public async Task<IActionResult> GetUserRequests(string userId)
    {
        var requests = await _requestsService.GetUserRequests(userId);
        return Ok(new { requests });
    }

    /// <summary>
    /// Cancel a request
    /// </summary>
    [HttpPost("{id}/cancel")]
    public async Task<IActionResult> CancelRequest(string id, [FromBody] Dictionary<string, string> body)
    {
        if (!body.TryGetValue("userId", out var userId))
        {
            return BadRequest("User ID is required");
        }

        var request = await _requestsService.CancelRequest(id, userId);
        if (request == null)
        {
            return NotFound(new { error = "Request not found" });
        }

        return Ok(request);
    }

    /// <summary>
    /// Complete a request
    /// </summary>
    [HttpPost("{id}/complete")]
    public async Task<IActionResult> CompleteRequest(string id, [FromBody] Dictionary<string, string> body)
    {
        body.TryGetValue("userId", out var userId);
        body.TryGetValue("providerId", out var providerId);
        body.TryGetValue("userRole", out var userRole);

        var finalUserId = userId ?? providerId;
        var finalUserRole = userRole ?? (providerId != null ? "provider" : "needy");

        if (string.IsNullOrEmpty(finalUserId))
        {
            return BadRequest(new
            {
                error = "User ID is required",
                message = "Please provide userId (or providerId for backward compatibility)"
            });
        }

        try
        {
            var request = await _requestsService.CompleteRequest(id, finalUserId, finalUserRole);
            
            if (request == null)
            {
                return NotFound(new { error = "Request not found" });
            }

            return Ok(new
            {
                success = true,
                request = new
                {
                    request.Id,
                    request.Status,
                    completedAt = request.UpdatedAt
                },
                message = "Request completed successfully. Moved to history."
            });
        }
        catch (Exception ex)
        {
            return BadRequest(new
            {
                error = ex.Message,
                message = ex.Message
            });
        }
    }

    // Quote endpoints

    /// <summary>
    /// Create a quote for a fuel request
    /// </summary>
    [HttpPost("quotes/create")]
    public async Task<IActionResult> CreateQuote([FromBody] CreateQuoteDto dto)
    {
        if (string.IsNullOrEmpty(dto.ProviderId))
        {
            return BadRequest(new
            {
                error = "Provider ID is required",
                message = "Please provide a valid provider ID"
            });
        }

        if (string.IsNullOrEmpty(dto.RequestId))
        {
            return BadRequest(new
            {
                error = "Request ID is required",
                message = "Please provide a valid request ID"
            });
        }

        Console.WriteLine($"💰 Creating quote: requestId={dto.RequestId}, providerId={dto.ProviderId}, price={dto.Price}");

        try
        {
            var quote = await _requestsService.CreateQuote(
                dto.RequestId,
                dto.ProviderId,
                dto.Price,
                dto.Currency,
                dto.EstimatedDeliveryTime,
                dto.Message);

            Console.WriteLine($"✅ Quote created successfully: {quote.Id}");

            return Ok(new
            {
                quote.Id,
                quote.RequestId,
                quote.ProviderId,
                quote.Price,
                quote.Currency,
                quote.EstimatedDeliveryTime,
                quote.Message,
                quote.Status,
                quote.CreatedAt,
                quote.UpdatedAt,
                successMessage = "Quote sent successfully to needy user"
            });
        }
        catch (Exception ex)
        {
            Console.WriteLine($"❌ Error creating quote: {ex.Message}");
            return BadRequest(new
            {
                error = ex.Message,
                message = ex.Message
            });
        }
    }

    /// <summary>
    /// Get quotes for a request
    /// </summary>
    [HttpGet("quotes/request/{requestId}")]
    public async Task<IActionResult> GetQuotesForRequest(string requestId)
    {
        var quotes = await _requestsService.GetQuotesForRequest(requestId);
        return Ok(new
        {
            quotes,
            count = quotes.Count
        });
    }

    /// <summary>
    /// Get quotes for a needy user
    /// </summary>
    [HttpGet("quotes/needy/{needyId}")]
    public async Task<IActionResult> GetQuotesForNeedy(string needyId)
    {
        var quotes = await _requestsService.GetQuotesForNeedy(needyId);
        return Ok(new
        {
            quotes,
            count = quotes.Count
        });
    }

    /// <summary>
    /// Get quotes for a provider
    /// </summary>
    [HttpGet("quotes/provider/{providerId}")]
    public async Task<IActionResult> GetQuotesForProvider(string providerId)
    {
        var quotes = await _requestsService.GetQuotesForProvider(providerId);
        return Ok(new
        {
            quotes,
            count = quotes.Count
        });
    }

    /// <summary>
    /// Accept a quote
    /// </summary>
    [HttpPost("quotes/{quoteId}/accept")]
    public async Task<IActionResult> AcceptQuote(string quoteId, [FromBody] AcceptQuoteDto dto)
    {
        try
        {
            Console.WriteLine($"✅ Accepting quote: quoteId={quoteId}, needyId={dto.NeedyId}");

            if (string.IsNullOrWhiteSpace(dto.NeedyId))
            {
                Console.WriteLine("❌ Needy ID is missing or empty");
                return BadRequest("Needy ID is required");
            }

            var quote = await _requestsService.AcceptQuote(quoteId, dto.NeedyId);
            
            if (quote == null)
            {
                return NotFound(new { error = "Quote not found" });
            }
            
            Console.WriteLine($"✅ Quote accepted successfully: {quote.Id}");

            var request = await _requestsService.GetRequest(quote.RequestId);
            if (request == null)
            {
                Console.WriteLine($"❌ Request {quote.RequestId} not found after accepting quote");
                return Ok(new
                {
                    quote.Id,
                    quote.Status,
                    request = (object?)null,
                    warning = "Request not found",
                    message = "Quote accepted successfully, but request details could not be retrieved."
                });
            }

            Console.WriteLine($"✅ Request retrieved: id={request.Id}, status={request.Status}");

            return Ok(new
            {
                quote.Id,
                quote.Status,
                quote.RequestId,
                request = new
                {
                    request.Id,
                    request.Status,
                    request.NeedyId,
                    request.AcceptedBy
                },
                message = "Quote accepted successfully. Request is now hidden from other providers."
            });
        }
        catch (Exception ex)
        {
            Console.WriteLine($"❌ Error accepting quote: {ex.Message}");
            return BadRequest(ex.Message);
        }
    }

    /// <summary>
    /// Reject a quote
    /// </summary>
    [HttpPost("quotes/{quoteId}/reject")]
    public async Task<IActionResult> RejectQuote(string quoteId, [FromBody] Dictionary<string, string> body)
    {
        if (!body.TryGetValue("needyId", out var needyId))
        {
            return BadRequest("Needy ID is required");
        }

        var quote = await _requestsService.RejectQuote(quoteId, needyId);
        if (quote == null)
        {
            return NotFound(new { error = "Quote not found" });
        }

        return Ok(new
        {
            quote.Id,
            quote.Status,
            message = "Quote rejected"
        });
    }

    /// <summary>
    /// Get a specific quote
    /// </summary>
    [HttpGet("quotes/{quoteId}")]
    public async Task<IActionResult> GetQuote(string quoteId)
    {
        var quote = await _requestsService.GetQuote(quoteId);
        if (quote == null)
        {
            return NotFound(new { error = "Quote not found" });
        }

        return Ok(quote);
    }
}

